src/openrct2/core/FileIndex.hpp

```diff
@@ -178,32 +179,88 @@ class FileIndex
         return ScanResult(stats, files);
     }
 
-    std::vector<TItem> Build(const ScanResult &scanResult) const
+    void BuildRange(const ScanResult &scanResult,
+                    size_t rangeStart,
+                    size_t rangeEnd,
+                    std::vector<TItem>& items,
+                    std::mutex& printLock) const
     {
-        std::vector<TItem> items;
-        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
-
-        auto startTime = std::chrono::high_resolution_clock::now();
-        // Start at 1, so that we can reach 100% completion status
-        size_t i = 1;
-        for (auto filePath : scanResult.Files)
+        items.reserve(rangeEnd - rangeStart);
+        for (size_t i = rangeStart; i < rangeEnd; i++)
         {
-            Console::WriteFormat("File %5d of %d, done %3d%%\r", i, scanResult.Files.size(), i * 100 / scanResult.Files.size());
-            i++;
-            log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            const auto& filePath = scanResult.Files.at(i);
+
+            if (_log_levels[DIAGNOSTIC_LEVEL_VERBOSE])
+            {
+                std::unique_lock<std::mutex> lock(printLock);
+                log_verbose("FileIndex:Indexing '%s'", filePath.c_str());
+            }
+
             auto item = Create(filePath);
             if (std::get<0>(item))
             {
                 items.push_back(std::get<1>(item));
             }
         }
+    }
+
+    std::vector<TItem> Build(const ScanResult &scanResult) const
+    {
+        std::vector<TItem> allItems;
+        Console::WriteLine("Building %s (%zu items)", _name.c_str(), scanResult.Files.size());
+
+        auto startTime = std::chrono::high_resolution_clock::now();
+
+        const size_t totalCount = scanResult.Files.size();
+        if (totalCount > 0)
+        {
+            JobPool jobPool;
+            std::mutex printLock; // For verbose prints.
 
-        WriteIndexFile(scanResult.Stats, items);
+            std::vector<std::vector<TItem>> containers;
+
+            size_t stepSize = std::thread::hardware_concurrency();
+            size_t numTasks = totalCount / stepSize;
+            size_t taskGroup = 0;
+            numTasks += (totalCount % stepSize == 0 ? 0 : 1);
```

Comment:
I don't get the logic behind these calculations. The more threads a system has, the fewer jobs it'll create. It should be exactly the opposite.
1 thread => 1 job for every object
2 threads => n/2 jobs that scan 2 objects each
4 threads => n/4 jobs that scan 4 objects each
etc.

I'd just pick a magic number (define it somewhere) to scan X objects per job (so that you don't have the overhead of calling the tasks and locking/unlocking mutexes as much). The amount of jobs will be the same, not matter how many cores a CPU has.